8278302: [s390] Implement fast-path for ASCII-compatible CharsetEncoders

RealLucy opened the pull request:
This pull request contains the s390x variant of the fast-path for ASCII-compatible CharsetEncoders. 

The intrinsic implementation is pretty much similar to that of PPC64. Compared to the previously existing intrinsic implementation, changes to program logic are minimal. There is a new match rule for ascii strings and the bit mask used to detect invalid bits in the string characters is set depending on the target charset: 0xff00 for ISO, 0xff80 for ascii.

**Request:** SAP does not maintain a full build and test infrastructure for s390x anymore. Could somebody please test this PR on s390x? At least build and tier1 tests, the more the better. Thank you!

With my private testing (some microbenchmarks, some SPEC benchmarks) I could not find any issues. With the fast-path active I see a 2x, sometime 2.5x, improvement of CharsetEncodeDecode.encode. 

**Without the fast-path patch:**
```
CharsetEncodeDecode.encode   16384        UTF-8  avgt   30  16.350 ± 0.067  us/op
CharsetEncodeDecode.encode   16384         BIG5  avgt   30  15.972 ± 0.136  us/op
CharsetEncodeDecode.encode   16384  ISO-8859-15  avgt   30  14.372 ± 0.110  us/op
CharsetEncodeDecode.encode   16384        ASCII  avgt   30  14.287 ± 0.051  us/op
```

**With the fast-path patch:**
```
CharsetEncodeDecode.encode   16384        UTF-8  avgt   30   6.833 ± 0.064  us/op
CharsetEncodeDecode.encode   16384         BIG5  avgt   30   8.195 ± 0.059  us/op
CharsetEncodeDecode.encode   16384  ISO-8859-15  avgt   30   6.387 ± 0.149  us/op
CharsetEncodeDecode.encode   16384        ASCII  avgt   30   6.321 ± 0.049  us/op
```

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8278302](https://bugs.openjdk.java.net/browse/JDK-8278302): [s390] Implement fast-path for ASCII-compatible CharsetEncoders


### Reviewers
 * [Martin Doerr](https://openjdk.java.net/census#mdoerr) (@TheRealMDoerr - **Reviewer**) ⚠️ Review applies to 4326bf93de4dd76edd78cee044e2a2b44a2822d2
 * @backwaterred (no known github.com user name / role)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6738/head:pull/6738` \
`$ git checkout pull/6738`

Update a local copy of the PR: \
`$ git checkout pull/6738` \
`$ git pull https://git.openjdk.java.net/jdk pull/6738/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6738`

View PR using the GUI difftool: \
`$ git pr show -t 6738`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6738.diff">https://git.openjdk.java.net/jdk/pull/6738.diff</a>

</details>

RealLucy commented:
Had already contacted Tim Ellison (RedHat) before Christmas. He promised to point an engineer to this PR.

backwaterred commented:
@RealLucy, sorry to keep you waiting. I am currently running tier1-tests on our s390x system. I'll edit this comment with the results when it completes.

---

RealLucy commented:
@backwaterred Thanks for working on the tests.
At first glance, the failure you are seeing doesn't immediately suggest it's my changes that are causing it. Do I understand the text "INTERPRETED" correctly if I assume the test was running without JIT compilation?

backwaterred commented:
I am not very familiar with it, but taking a quick look at [the test](https://github.com/openjdk/jdk/blob/master/test/hotspot/jtreg/compiler/unsafe/UnsafeCopyMemory.java) I think you are right. Since the failure occurred after only a few tests in the run, I believe the failure occurred in the interpreted section.

backwaterred commented:
If you prefer it, the test result files are also accessible here:
https://ibm.box.com/s/n1l6e1vpncpsqqvzgbm3czpulmibu8oz

RealLucy commented:
Thank you for testing! The results don't look good, though. I'll have to find out what the test expects and why my code doesn't deliver the expected. Will get back to you asap. 
Lutz

RealLucy commented:
@backwaterred I have updated the PR with a commit fixing the issue found during testing. The only code change in this commit is in line #80, where I removed the "int", turning the declaration/definition into an assignment. All else is just code comments. 

I would appreciate very much if you could run the tests once again. There is one more thing: could you, or somebody else in your team, please review the change? Thx.

backwaterred commented:
@RealLucy I am happy to re-run the tests and review your changes. I believe you will need a second reviewer as well. Hopefully we can encourage someone from the mailing lists to come check out the PR, and I will also ask around.

backwaterred commented:
The most recent run is looking good. Nicely done.

I will review your changes ASAP. Should be in the next day or so, depending on how long it takes me to understand them :-P (thank you for the extra comments).

RealLucy commented:
@backwaterred Yes, for hotspot changes, two reviews are required. At least one review must be from a "real" reviewer.

RealLucy commented:
Let me aid understanding with a few additional comments. 

The implementation of C2_MacroAssembler::string_compress() was pre-existing for the plain compress function and for encoding ISO (8-bit character) arrays. What I had to add was another mask value to support 7-bit ascii characters. The encoding logic remained the same. 

In the s390.ad file, a new match rule was added for 7-bit ascii characters. This rule is, with the exception of the predicate, identical to the one for ISO arrays.

cl4es reviewed (commented):
I can't really review this, though it looks like a pretty straightforward port of the x86 patch. Nice work!

While hotspot requires 2 reviewers for patches, this is only touching s390-specific code. I'd then expect this to be governed more directly by the s390 project/maintainers, who should be free to use the slightly more relaxed default OpenJDK rules (only one reviewer required) if they so choose.

RealLucy commented:
Thank you for your comments, Claes!
As SAP has passed on responsibility for the s390 port, I would like to receive a "lead" review from the new maintainers. Then, I could recruit a "second opinion" from my direct colleagues.

TheRealMDoerr commented on src/hotspot/cpu/s390/c2_MacroAssembler_s390.cpp:
```diff
@@ -46,20 +46,41 @@
+      BLOCK_COMMENT("encode_iso_array {");
+    }
   } else {
     BLOCK_COMMENT("string_compress {");
```
Maybe assert(!toASCII, "unsupported") ?

RealLucy replied:
Hmm,
could be done. On the other hand, calls are exclusively from ad-file match rules. But ok,  I'll create another commit.

TheRealMDoerr reviewed (commented):
Looks good to me. Should get tested.

RealLucy commented:
Tests are currently being performed by @backwaterred 
Expecting results soon.

backwaterred commented:
The results are in @RealLucy. All tests passing!

RealLucy commented:
Wonderful! 
Thank you for your effort.

RealLucy commented:
Any other potential Reviewer out there? Thanks!

backwaterred commented:
These changes look good. Thanks again.

RealLucy commented:
/integrate
